### PR TITLE
feat: WebFrameMain.collectJavaScriptCallStack()

### PR DESCRIPTION
**Pull request by samuelmaddock**
#### Description of Change

Adds `WebFrameMain.collectJavaScriptCallStack()` to allow accessing the JavaScript call stack of unresponsive renderers. This is available in Chromium as part of the [Crash Reporting API.](https://wicg.github.io/crash-reporting/)

With this API, a dialog shown when a renderer becomes unresponsive could have an option to copy the stack trace. This would help application developers figure out why their website might be stuck on long-running JavaScript.

I've marked the API as experimental given that the proposal is still a draft.

-----

Here's an example of the output from a [fiddle gist](https://gist.github.com/dfddc4c3e4459f097ca7f9bb35218d88):
```
Renderer unresponsive
 
    at startInfiniteLoop (<anonymous>:5:5)
    at HTMLHtmlElement.clickHandler (<anonymous>:11:5)
```

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Added `WebFrameMain.collectJavaScriptCallStack()` for accessing the JavaScript call stack of unresponsive renderers.

### CODE REVIEW

File: docs/api/web-frame-main.md
Comment (nornagon-openai) [inline docs/api/web-frame-main.md:234]: this is a strange API name for this shape. Does this only give you a call stack if the document is unresponsive? what happens if you access it while the document is _not_ unresponsive? If it's required that the document be unresponsive for this to work, should this instead be passed as part of the unresponsive event? How should a caller know whether it's available or not?

Comment (nornagon) [inline docs/api/web-frame-main.md:234]: this is a strange API name for this shape. Does this only give you a call stack if the document is unresponsive? what happens if you access it while the document is _not_ unresponsive? If it's required that the document be unresponsive for this to work, should this instead be passed as part of the unresponsive event? How should a caller know whether it's available or not?

Comment (samuelmaddock) [reply docs/api/web-frame-main.md:234]: In Chromium, when a WebContents becomes unresponsive, it triggers a Mojo IPC to collect the JS call stack via [content::RenderProcessHostImpl::InterruptJavaScriptIsolateAndCollectCallStack()](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/web_contents/web_contents_impl.cc;l=9385-9395;drc=1d694679493c7b2f7b9df00e967b4f8699321093). There's no observer event to indicate when the collection has completed unfortunately. Currently it's read when a crash report is being generated.

If we were to patch Blink or propose a change upstream, a better API might be
```js
webContents.on('unresponsive-js-call-stack-updated', ({ callStack }) => {
  console.log('Updated call stack', callStack);
})
```

In this feature's current state, this will only ever be accessible if the Chrome feature is enabled with the `Document-Policy` header also set on the unresponsive webpage.

Given the draft status and early implementation, I was leaning towards this being an experimental API for now until something cleaner is available. I'd also be ok with just not documenting this API yet until the standard is finalized.

Comment (samuelmaddock) [reply docs/api/web-frame-main.md:234]: Alternative idea: once the call stack is received, emit `unresponsive` again, now with the call stack details. Again though, this would require a chromium patch.

Comment (deepak1556) [reply docs/api/web-frame-main.md:234]: The interface used for capturing the stack https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/public/mojom/call_stack_generator/call_stack_generator.mojom does not specify the requirement of document being in unresponsive state, it seems to be valid to call at any point in time of main thread execution.

Option 1) Could we get a handle to this interface exposed in the public API of `RenderProcessHost`, so that we can listen on the interface response to know when stack collection is compelete. This would make the `WebFrameMain` api more generic.

Option 2) If the feature requirements are met, then `unresponsive` event is emitted with stack details only after the stack collection is complete. This does require changes in upstream when it emits `WebContentsObserver::OnRendererUnresponsive` but takes away the need for a user to call into additional api when they have already signaled that they need stack details.

Comment (itsananderson) [reply docs/api/web-frame-main.md:234]: > In this feature's current state, this will only ever be accessible if the Chrome feature is enabled with the Document-Policy header also set on the unresponsive webpage.

It would be nice for this to not be required, though it's probably not worth maintaining an extra patch since it seems fairly easy to work around.

Comment (samuelmaddock) [reply docs/api/web-frame-main.md:234]: > Could we get a handle to this interface exposed in the public API of RenderProcessHost, so that we can listen on the interface response to know when stack collection is compelete.

Once I have more time to focus on this, I'll give this a try.

> If the feature requirements are met, then unresponsive event is emitted with stack details only after the stack collection is complete.

I believe there are cases where it's unsuccessful in capturing the call stack. Would need to measure if that adds additional time to the round-trip IPCs. That would be convenient if it could be included with the event!

> It would be nice for this to not be required, though it's probably not worth maintaining an extra patch since it seems fairly easy to work around.

To clarify, this feature will eventually be enabled by default presumably once the draft and implementation upstream has completed.

-----

btw, I've had luck using this in our application by enabling the necessary feature flags. It sends the call stacks to the reporting server. Would recommend trying that route at least before this feature is available!

Comment (nornagon) [reply docs/api/web-frame-main.md:234]: If we have the APIs to be able to collect a call stack at any time, let’s expose that!

if we can *additionally* hook into logic that chrome uses to collect call stacks when unresponsive, that seems useful too, but we already have logic for detecting unresponsive frames and if we had an API for grabbing a stack then clients could just use it in their event handlers.

**Comment by samuelmaddock** [comment]
> If we have the APIs to be able to collect a call stack at any time, let’s expose that!

Updated the API to do just that. cc @nornagon 

There are still some requirements to be able to collect the call stack:
- JavaScript must be running. Otherwise the [RequestInterrupt](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/controller/javascript_call_stack_collector.cc;l=125;drc=86645dfc93a07bfd0b49b78aa06b292c8149a075) never resolves.
- The webpage owner needs to enable the document-policy header defined by the crash reporting API
- The origin trial needs to be enabled or the `DocumentPolicyIncludeJSCallStacksInCrashReports` chromium flag.

File: docs/api/web-frame-main.md
Comment (deepak1556) [inline docs/api/web-frame-main.md:148]: > If no JavaScript runs, the promise will never resolve

Just curious, what is the better way to determine this case ? would checking for [current stack position](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/utils/utils.cc;l=222) against the threads stack base address help ?

File: spec/index.js
Comment (deepak1556) [inline spec/index.js:39]: Any reason to not enable this feature by default `DocumentPolicyIncludeJSCallStacksInCrashReports` in the runtime ? Users are still required to opt-in respective documents via the policy header, just wondering if there is necessity to control the feature behind two runtime feature flags. Maybe in our case just the document policy is sufficient, thoughts ?

Comment (samuelmaddock) [reply spec/index.js:39]: I didn't consider this given the early API availability in Chromium. It's estimated to be in origin trial until Chromium 132 and there hasn't been significant feedback by standards committees based on information provided in [the Chrome Status page.](https://chromestatus.com/feature/4731248572628992)

Assuming it's fully rolled out in Chromium 133, we could expect to not require the flag by Feb 4, 2025.

> Maybe in our case just the document policy is sufficient, thoughts?

I'm not against it if folks are okay with this.

**Review by VerteDinde — APPROVED** [review]
From a code implementation side, this looks good to me - the diff is relatively small and low-risk 👍 Approved pending API working group approval

File: spec/index.js
Comment (deepak1556) [reply spec/index.js:39]: Sounds good, we can revisit this again.

File: docs/api/web-frame-main.md
Comment (samuelmaddock) [reply docs/api/web-frame-main.md:148]: > would checking for [current stack position](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/utils/utils.cc;l=222) against the threads stack base address help ?

I think this would be best left as feedback (or a CL) upstream so it can be fixed there. We should mention this in https://issues.chromium.org/issues/40268201

> Naively, would providing a timeout argument be helpful to provide users with an escape hatch?

If a fix cannot be applied upstream, I think we should consider this. My thinking is we merge this as experimental for now and revisit after seeing results from usage.

**Review by erickzhao — APPROVED** [review]
API LGTM

**Review by itsananderson — APPROVED** [review]
API LGTM

**Comment by deepak1556** [comment]
PR should be good to merge, can you rebase on latest `main`

**Review by deepak1556 — APPROVED** [review]

File: spec/index.js
Comment (nikwen) [reply spec/index.js:39]: There's more discussion about it here: https://github.com/electron/electron/issues/45356#issuecomment-2634879409

**Pull request merged**
